### PR TITLE
157876050 Endpoint to get one asset-category 

### PR DESCRIPTION
**Pull request by purpose50**
#### What does this PR do?
get one asset category

#### Description of Task to be completed?
* create an endpoint to get asset category
* write a test for getting one category

#### How should this be manually tested?
* pull the branch
* run ``` pip install -r requrements.txt ```
* navigate to ```http://127.0.0.1:5000/api/v1/asset_categories/<string:id>``` on postman
* run ```pytest --setup-show``` to run the test

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
#157876050 An endpoint to get one asset category

#### Screenshots (if appropriate)
screen shot 2018-06-14 at 12 51 15 pm

screen shot 2018-06-14 at 12 50 30 pm

#### Questions:
N/A

### CODE REVIEW

File: api/views/asset_category.py [hunk 1]
```diff
@@ -30,3 +33,25 @@ def get(self):
...
             'data': data
         }
+
+
+@api.route('/asset-categories/<string:id>')
+class AssetCategoryResource(Resource):
+    @token_required
+    def get(self, id):
+        """
+        get a single asset categories
```
Comment (ibutiti) [inline api/views/asset_category.py]: Should be `category` not `categories`. Also capitalise first letter.

File: api/views/asset_category.py [hunk 2]
```diff
@@ -30,3 +33,25 @@ def get(self):
...
+        """
+        single_category = AssetCategory.get(id)
+        if not single_category or single_category.deleted is True:
```
Comment (ibutiti) [inline api/views/asset_category.py]: The `is True` is unnecessary.

File: api/views/asset_category.py [hunk 3]
```diff
@@ -30,3 +33,25 @@ def get(self):
...
+            raise ValidationError(dict(message='Asset category not found'), 404)
+
+        asset_category_schema = AssetCategorySchema()
+        attributes_schema = AttributeSchema(
+            many=True, exclude=['choices', 'id', 'deleted'])
+
+        return {
```
Comment (ibutiti) [inline api/views/asset_category.py]: You need to serialize the entire object to json. Use jsonify.

File: tests/conftest.py [hunk 4]
```diff
@@ -1,7 +1,7 @@
 import os
-
 import pytest
 
+
```
Comment (ibutiti) [inline tests/conftest.py]: Please refer to the import statement arrangement recommended by @brayoh

Comment (ibutiti) [reply tests/conftest.py]: Also, docstrings

File: api/views/asset_category.py [hunk 5]
```diff
@@ -4,6 +4,9 @@
 from api.models import AssetCategory
 from main import api
 from api.middlewares.token_required import token_required
+from api.middlewares.base_validator import ValidationError
```
Comment (ibutiti) [inline api/views/asset_category.py]: Docstrings

File: tests/test_asset_categories_endpoints.py [hunk 6]
```diff
@@ -19,3 +19,31 @@ def test_asset_categories_stats_endpoint(self, client, new_asset_category, init_
...
+
+        response = client.get('{}/asset-categories/{}'.format(
+          api_v1_base_url, asset_category.id), headers=auth_header)
+
+        response_json = json.loads(response.data.decode(CHARSET))
+
+        assert response.status_code == 200
+        assert response_json['name'] == 'TestLaptop'
+        assert type(response_json['customAttributes']) == list
+
```
Comment (ibutiti) [inline tests/test_asset_categories_endpoints.py]: Also test for edge case of symbols, invalid strings etc as the id argument supplied in url so we see if the validation i'd recommended catches them.

File: api/views/asset_category.py [hunk 7]
```diff
@@ -30,3 +33,25 @@ def get(self):
...
+
+
+@api.route('/asset-categories/<string:id>')
+class AssetCategoryResource(Resource):
+    @token_required
+    def get(self, id):
+        """
+        get a single asset categories
+        """
+        single_category = AssetCategory.get(id)
```
Comment (ibutiti) [inline api/views/asset_category.py]: Since we know how the ids look, we can validate the id before hitting the db with a query.

File: api/views/asset_category.py (hunk 1 above)
Comment (purpose50) [reply api/views/asset_category.py]: Thanks @ibutiti  on it

File: api/views/asset_category.py (hunk 2 above)
Comment (purpose50) [reply api/views/asset_category.py]: Yes, thought I removed it already, on it.

File: api/views/asset_category.py (hunk 7 above)
Comment (purpose50) [reply api/views/asset_category.py]: Thanks @ibutiti , the reason why I didn't validate the `id` is that whatever the user supply as the `id`, as far as it is not in the id field of the table it will return `asset category not found` and the application will not break.

I also thought of what I should be validating for since the `id` we are expecting is a `string`, even if the user supplies an `integer`, the application won't break and the user will get an error that the `asset category is not found`. nevertheless, I'll try and `validate the id` to ensure it is a `string` so we don't leave any stone untouched.

Comment (purpose50) [reply api/views/asset_category.py]: Hi @ibutiti made a new discovery. because the endpoint already defined the type of the id, it automatically converts it to  that type..for example:
this is the endpoint  `/asset-categories/<string:id>`. from the endpoint we are obviously expecting the id to be a string, so any id supplied from postman is automatically converted to a string..so all  of these will be automatically converted to a string:
`/asset-categories/1`
`/asset-categories/222`
`/asset-categories/@@@@`
`/asset-categories/0928ujj`
`/asset-categories/%%%%%`

now if the `id` suplied is not found in the database then the user gets a response that the `asset category is not found`, it doesnt break the app, it just returns a message that the `asset category with that id is not found` even though the `id` is just a set of random characters

**Comment by brayoh** [comment]
@purpose50 Kindly `squash` your commits into 1

File: api/views/asset_category.py (hunk 3 above)
Comment (Dheavyman) [reply api/views/asset_category.py]: @ibutiti I believe flask-restplus has already converted the object to json response with the return. The header already has the application/json added too. From flask-restplus doc, "Flask-RESTPlus understands multiple kinds of return values from view methods. Similar to Flask, you can return any iterable and it will be converted into a response, including raw Flask response objects."

File: api/views/asset_category.py (hunk 7 above)
Comment (ibutiti) [reply api/views/asset_category.py]: Agreed @purpose50. However, since we know that the push_id uses these characters: ```PUSH_CHARS = ('-0123456789'
                  'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
                  '_abcdefghijklmnopqrstuvwxyz')```
we can do a simple regex to check that the characters in the id only appear here. This operation will be a lot cheaper than a DB query to let us know that the asset category was not found.

File: api/views/asset_category.py (hunk 3 above)
Comment (ibutiti) [reply api/views/asset_category.py]: ok @Dheavyman

File: tests/test_asset_categories_endpoints.py [hunk 8]
```diff
@@ -19,3 +20,31 @@ def test_asset_categories_stats_endpoint(self, client, new_asset_category, init_
         assert len(response_json['data']) is not 0
         assert response_json['data'][0]['name'] == 'Laptop'
         new_asset_category.delete()
+
+    """
+    Tests that a single asset category can be retrieved
```
Comment (ibutiti) [inline tests/test_asset_categories_endpoints.py]: Please put this docstring inside the test

Comment (purpose50) [reply tests/test_asset_categories_endpoints.py]: sure, thanks

File: api/views/asset_category.py (hunk 7 above)
Comment (SEUNAGBEYE) [reply api/views/asset_category.py]: @ibutiti I concur, but we still need to verify if the item is in the `db`

```
if not (re.match('-[a-zA-Z0-9]', id) and Asset_category.get(id)):
                        raise Validation({'message': 'Asset Category not found'})
```

Comment (SEUNAGBEYE) [reply api/views/asset_category.py]: @purpose50 I suggest you put that in a function somewhere in the `utilities` folder or somewhere else because we would be validating the `ids` in several places and it won't be idea repeating the validation

File: api/views/asset_category.py [hunk 9]
```diff
@@ -30,3 +36,29 @@ def get(self):
...
+
+
+@api.route('/asset-categories/<string:id>')
+class AssetCategoryResource(Resource):
+    @token_required
+    def get(self, id):
+        """
+        get a single asset category
+        """
+        if not (re.match('-[a-zA-Z0-9]', id)):
```
Comment (Dheavyman) [inline api/views/asset_category.py]: Please abstract this to a function in utilities folder so that it can be used in other places required.

File: api/views/asset_category.py [hunk 10]
```diff
@@ -30,3 +37,29 @@ def get(self):
...
+
+        asset_category_schema = AssetCategorySchema()
+        attributes_schema = AttributeSchema(
+            many=True, exclude=['choices', 'id', 'deleted'])
+
+        return {
+           'name': single_category.name,
+           'customAttributes': attributes_schema.dump(
+               single_category.attributes).data
+        }, 200
```
Comment (Dheavyman) [inline api/views/asset_category.py]: The response object should have status and data properties, with the data holding the information as described above. The status should either be success or error depending on the result of the request.

File: tests/conftest.py [hunk 11]
```diff
@@ -64,6 +65,12 @@ def new_asset_category(app):
     return asset_category
 
 
+"""
+The module scope is used here to prevent a test module data leaking into
+another
+"""
+
+
```
Comment (brayoh) [inline tests/conftest.py]: I think it would be better if we have this doc string inside the test method instead
